Diagnostic.py

```diff
@@ -1,332 +1,352 @@
+    return Jacobian.vectorize(grad(Jacobian.vectorize(self.first_grad),self.params,v,retain_graph=True))
+
+  def _matvec(self,v):
+    v = np.expand_dims(v,axis=-1)
+    v = torch.tensor(v,device=self.device)
+    if self.transpose == True:
+      return self.JTVP(v).cpu().detach().numpy()
+    else:
+      return self.JVP(v).cpu().detach().numpy()
+
+  def trace(self,n_iter): ### n_iter for Hutchinson Stochastic Trace Estimators
+    trace_list = []
+    for _ in range(n_iter):
+      v = self.rand_vec()
+      trace_list.append(torch.matmul(v.T,self.JTVP(v)))
+    return torch.mean(torch.tensor(trace_list))
+  
+  def eigen_pair(self,n_eigen,which):  #### returns [eig_val, eig_vec]
+    
+    #### WHICH ################
+    #‘LM’ : largest magnitude
+    #‘SM’ : smallest magnitude
+    #‘LR’ : largest real part
+    #‘SR’ : smallest real part
+    #‘LI’ : largest imaginary part
+    #‘SI’ : smallest imaginary part
+    
+    if self.transpose == True:
+      return eigsh(self,which = which,k=n_eigen)
+    else:
+      return eigs(self,which = which,k=n_eigen)
+
+  '''random vector from Rademacher distribution'''
+
+  def rand_vec(self):
+    v = torch.cat([torch.randint_like(p.data, high=2,device=self.device).flatten()\
+          for p in self.params]).unsqueeze(-1)
+    v[v==0.]=-1. 
+    return v/np.sqrt(len(v))
+
+  '''Orthogonilise: Gram-Schmidt process '''
+
+  @staticmethod
+  def ort(w,list1):
+    for v in list1:
+      w = w - torch.matmul(w.T,v)*v
+      w = w/torch.norm(w)
+    return w
+
+  '''Lancazos algorithm'''
+
+  def lanczos(self,n_iter):  ### Returns V basis and T tridiagonal matrix
+                              
+    v       = self.rand_vec()
+    w_prime = self.JTVP(v)
+    alpha   = torch.matmul(w_prime.T,v)
+    w       = w_prime - alpha*v  ## projection of w_prime on v
+    v_list      = [v]
+    w_list      = [w]
+    alpha_list  = [alpha]
+    beta_list   = []
+
+    V = torch.zeros((self.shape[0],n_iter))
+    T = torch.zeros((n_iter,n_iter))
+    T[0,0] = alpha
+    V[:,0] = v.squeeze(-1)
+
+    for i in range(n_iter-1):
+
+      betta = torch.norm(w_list[-1])
+      beta_list.append(betta)
+
+      T[i,i+1] = betta
+      T[i+1,i] = betta
+
+      if betta !=0:
+        v = w_list[-1]/betta
+        v = Jacobian.ort(v,v_list)
+      else:
+        v = Jacobian.ort(self.rand_vec(),v_list)
+
+      v_list.append(v)
+      V[:,i+1] = v.squeeze(-1)
+
+      w_prime = self.JTVP(v_list[-1])
+      alpha = torch.matmul(w_prime.T,v_list[-1])
+      alpha_list.append(alpha)
+      T[i+1,i+1] = alpha
+
+      w = w_prime-alpha_list[-1]*v_list[-1]-beta_list[-1]*v_list[-2]
+      w_list.append(w)
+    return V, T
+  
+def min_max(X): #X 2D array
+  x1_gen = X[:,0]
+  x2_gen = X[:,1]
+  return np.min(x1_gen),np.max(x2_gen),\
+         np.min(x2_gen),np.max(x2_gen)
+
+'''vis function plots GAN 2D generated vs real points, with background coloured by discriminator output'''
+
+def vis(G,D,X_train_save,z_fixed,device):
```

Comment:
It is had to understand what `G` and `D` mean. Maybe you can better names for your variables.
Also local variables and parameters are by convention lowercase